futures-util/src/stream/mapped_futures/mod.rs

```diff
@@ -0,0 +1,473 @@
+    ///
+    /// This method adds the given future to the set. This method will not
+    /// call [`poll`](core::future::Future::poll) on the submitted future. The caller must
+    /// ensure that [`MappedFutures::poll_next`](Stream::poll_next) is called
+    /// in order to receive wake-up notifications for the given future.
+    /// Returns true if another future was ma
+    pub fn replace(&mut self, key: K, future: Fut) -> Option<Fut>
+    where
+        Fut: Unpin,
+    {
+        let replacing = self.remove(&key);
+        let task = self.inner.push(key, future);
+        // self.hash_set.insert(task.into());
+        self.set().insert(task.into());
+        replacing
+    }
+
+    /// Remove a future from the set, dropping it.
+    ///
+    /// Returns true if a future was cancelled.
+    pub fn cancel(&mut self, key: &K) -> bool {
+        // if let Some(task) = self.hash_set.get(key) {
+        if let Some(task) = self.set().take(key) {
+            unsafe {
+                if let Some(_) = (*task.future.get()).take() {
+                    self.inner.unlink(Arc::as_ptr(&task.inner));
+                    return true;
+                }
+            }
+        }
+        false
+    }
+
+    /// Remove a future from the set and return it.
+    pub fn remove(&mut self, key: &K) -> Option<Fut>
+    where
+        Fut: Unpin,
+    {
+        // if let Some(task) = self.hash_set.get(key) {
+        if let Some(task) = self.set().take(key) {
+            unsafe {
+                let fut = (*task.future.get()).take().unwrap();
+                self.inner.unlink(Arc::as_ptr(&task.inner));
+                return Some(fut);
+            }
+        }
+        None
+    }
+
+    /// Returns `true` if the map contains a future for the specified key.
+    pub fn contains(&mut self, key: &K) -> bool {
+        self.set().contains(key)
+    }
+
+    /// Get a pinned mutable reference to the mapped future.
+    pub fn get_pin_mut(&mut self, key: &K) -> Option<Pin<&mut Fut>> {
+        // if let Some(task_ref) = self.hash_set.get(key) {
+        if let Some(task_ref) = self.set().get(key) {
+            unsafe {
+                if let Some(fut) = &mut *task_ref.inner.future.get() {
+                    return Some(Pin::new_unchecked(fut));
+                }
+            }
+        }
+        None
+    }
+
+    /// Get a pinned mutable reference to the mapped future.
+    pub fn get_mut(&mut self, key: &K) -> Option<&mut Fut>
+    where
+        Fut: Unpin,
+    {
+        // if let Some(task_ref) = self.hash_set.get(key) {
+        if let Some(task_ref) = self.set().get(key) {
+            unsafe {
+                if let Some(fut) = &mut *task_ref.inner.future.get() {
+                    return Some(fut);
+                }
+            }
+        }
+        None
+    }
+
+    /// Get a shared reference to the mapped future.
+    pub fn get(&mut self, key: &K) -> Option<&Fut>
+    where
+        Fut: Unpin,
+    {
```

Comment:
`Pin` always implements `Deref` (i.e., we can always safely get `&T` from `Pin<&T>`/`Pin<&mut T>`), so I think we can remove the `Unpin` bound here.